src/kv/KeyValueDB.h

```diff
@@ -226,77 +246,138 @@ class KeyValueDB {
   };
   typedef ceph::shared_ptr< WholeSpaceIteratorImpl > WholeSpaceIterator;
 
+  /// Iterator for column family which is created explicitly and not the default one.
+  class ColumnFamilyIteratorImpl {
+  public:
+    virtual int seek_to_first() = 0;
+    virtual int seek_to_last() = 0;
+    virtual int upper_bound(const std::string &after) = 0;
+    virtual int lower_bound(const std::string &to) = 0;
+    virtual bool valid() = 0;
+    virtual int next() = 0;
+    virtual int prev() = 0;
+    virtual std::string key() = 0;
+    virtual bufferlist value() = 0;
+    virtual bufferptr value_as_ptr() {
+      bufferlist bl = value();
+      if (bl.length()) {
+        return *bl.buffers().begin();
+      } else {
+        return bufferptr();
+      }
+    }
+    virtual int status() = 0;
+    virtual ~ColumnFamilyIteratorImpl() { }
+  };
+  typedef ceph::shared_ptr< ColumnFamilyIteratorImpl > ColumnFamilyIterator;
+
   class IteratorImpl : public GenericIteratorImpl {
-    const std::string prefix;
+    const std::string prefix;         //prefix or CF name
     WholeSpaceIterator generic_iter;
+    ColumnFamilyIterator cf_iter;
+    bool is_cf;                       //is it for a explicit column family
   public:
     IteratorImpl(const std::string &prefix, WholeSpaceIterator iter) :
-      prefix(prefix), generic_iter(iter) { }
+      prefix(prefix), generic_iter(iter), is_cf(false) { }
+    IteratorImpl(const std::string &cf_name, ColumnFamilyIterator iter) :
+      prefix(cf_name), cf_iter(iter), is_cf(true) { }
     ~IteratorImpl() override { }
 
     int seek_to_first() override {
-      return generic_iter->seek_to_first(prefix);
+      return is_cf ? cf_iter->seek_to_first() :
+                     generic_iter->seek_to_first(prefix);
     }
     int seek_to_last() {
-      return generic_iter->seek_to_last(prefix);
+      return is_cf ? cf_iter->seek_to_last() :
+                     generic_iter->seek_to_last(prefix);
     }
     int upper_bound(const std::string &after) override {
-      return generic_iter->upper_bound(prefix, after);
+      return is_cf ? cf_iter->upper_bound(after) :
+                     generic_iter->upper_bound(prefix, after);
     }
     int lower_bound(const std::string &to) override {
-      return generic_iter->lower_bound(prefix, to);
+      return is_cf ? cf_iter->lower_bound(to) :
+                     generic_iter->lower_bound(prefix, to);
     }
     bool valid() override {
-      if (!generic_iter->valid())
-	return false;
-      return generic_iter->raw_key_is_prefixed(prefix);
+      if (is_cf) {
+        return cf_iter->valid();
+      } else {
+        if (!generic_iter->valid())
+	  return false;
+        return generic_iter->raw_key_is_prefixed(prefix);
+      }
```

Comment:
May be we can reorganize this code better to check for invalid.

Reply:
do you have any suggestion?